# 异常
异常:在程序出错误的时候， Java 支持使用异常，将错误信息封装起来， 并让程序跳出正常的处理流程，交给异常处理部分去处理。
try{语句}catch(Exception ex){ex.printStackTrace()}

# 按照异常的继承关系分类
1、异常也是 Java 中的类
2、所有异常的父类: Throwable
3、两类异常: Error（内部不够啊，等品台应该解决的导致系统崩溃的错误） 和 Exception（和自己程序相关的，语法错误   ===》 主要关心Exception 类的错误） 
4、 看一下 Throwable 的继承关系 （两下Shift =》control+H）


# 按照处理方式不同分类
1、checked exception: 语法要求必须要用 try catch 或者 throws 语句处理 的异常
2、unchecked exception: 语法不要求一定要用 try catch 或者 throws 语句 处理的异常
3、Error 和 RuntimeException 是 unchecked exception 的父类。我们一般 使用 RuntimeException


# throw 和 throws
throws: 方法签名上抛出异常（异常类型）,多个异常用逗号分开
throw: 在代码里创建一个异常：throw new Exception("");

# Java 异常的传递
1、方法一个调一个的层层调用，组成了方法调用栈
2、Java 异常的的归宿:要么沿着方法调栈顺序一路抛，最终造成当前线 程出错退出，要么被 catch 住

# 自定义异常
异常最重要的信息:类型、错误信息、出错时的调用栈（不能定义）

# 异常传递不是凌波微步
1、异常代表程序出错，不要使用异常来做正常的程序跳转
2、错误的示例:使用“凌波微步”跳转到 catch 语句
3、异常不是凌波微步，异常的创建和处理很耗费资源

# try
try
{
    语句
} catch (Exception ex) 
{
    ex.printStackTrace()
} finally {
    不管try和catch语句里return，还是结束了什么，finally都会执行
}

// finally里改变catch里边return值，无效 （在finally里不要有return）
finally 在 return 新的值可以覆盖之前的值

System.exit(0)之后 finally 就不执行了   

# 常见异常
• NullPointerException
• IndexOutOfBoundsException 
• ClassCastException  （类强制转换的时候 【一个类型的引用，转换成另一个类型的引用】）
• ClassNotFoundException
• IOException
